hairpin support

salemsol opened the pull request.

salemsol committed:
doca-dpdk: Add support for forward mode port

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
doca-flow: Remove check for monitor

The can either create a monitor or not, in case
no monitor is specified then the flow shouldn't fail.
Fix it.

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
doca-flow: Change fwd port type

DPDK port id are unsigned, change the port type in doca
flow accordingly.

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
gw-port: Add support creating haripin queues

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
main: Add option for creating hairpin queues

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol commented on examples/vnet/gw_port.c:
```diff
@@ -44,10 +44,15 @@ static void gw_assert_link_status(int port_id)
 }
 
-int gw_init_port(int port_id, int nr_queues)
+int sf_start_dpdk_port(struct sf_port_cfg *port)
```
unique to the simple_fwd application

rbaryanai commented on examples/vnet/doca_dpdk.c:
```diff
@@ -1232,6 +1258,10 @@ doca_dpdk_create_pipe(struct doca_flow_pipe_cfg *cfg,
 		flow->item_entry[idx].item = &pl->flow.items[idx];
 	for (idx = 0; idx < MAX_ACTIONS; idx++)
 		flow->action_entry[idx].action = &pl->flow.actions[idx];
+    if (fwd && fwd->type == DOCA_FWD_PORT)
```
since there is ++ on number of actions this will not work when having many entries,

salemsol replied:
can we have more than one fwd per pipe?
if not then i don't understand how it's related, can you please explain ?

rbaryanai commented on examples/vnet/doca_dpdk.h:
```diff
@@ -148,6 +148,11 @@ struct doca_dpdk_action_l4_port_data {
 	struct rte_flow_action_set_tp l4port;
 };
 
+struct doca_dpdk_action_fwd_data {
```
why it is not part of the union?

salemsol replied:
it is. "doca_dpdk_action_fwd_data " is part of "doca_dpdk_action_data"
unless you mean to remove doca_dpdk_action_fwd_data  and have two actions inside doca_dpdk_action_data
1. rte_flow_action_queue 
2. rte_flow_action_port_id 

i thought it would be better to have as one action "fwd" and then specify the action (queue / port) accordingly.

rbaryanai commented on examples/vnet/doca_flow.c:
```diff
@@ -38,12 +38,11 @@ struct doca_flow_pipe_entry *doca_flow_pipe_add_entry(
-	return doca_dpdk_pipe_create_flow(pipe, pipe_queue, match, actions, mon,
-					  (fwd != NULL)?fwd:&pipe->fwd,
-                                           err);
+	return doca_dpdk_pipe_create_flow(pipe, pipe_queue, match, actions, NULL,
```
you removed the mon

salemsol replied:
will revert

salemsol replied:
but what if the user doesn't provide a mon then we will fail to add the entry, i think mon should be treated as fwd this is why i removed it from the if, what do you think ?

rbaryanai commented on examples/vnet/doca_flow.h:
```diff
@@ -193,7 +193,8 @@ struct doca_flow_fwd {
         struct port {
-                int id;
+                uint16_t id;
+                bool hairpin;
```
hairpin is not part of doca_flow
lets discuss

salemsol replied:
hairpin is just a bool that we set upon building the fwd.
setting this bool as true indicates that we are in nic mode then the "id" is the hairpin's queue index.
setting this bool as false indicates that we are in switchdev.

this is done in the application. can be changed i think

rbaryanai commented on examples/vnet/gw_port.c:
```diff
@@ -44,10 +44,15 @@ static void gw_assert_link_status(int port_id)
 }
 
-int gw_init_port(int port_id, int nr_queues)
+int sf_start_dpdk_port(struct sf_port_cfg *port)
```
not following why you changed the function name

salemsol replied:
because we start the port here we cann "rte_eth_dev_start".
after starting the port we need to bind the queues to the ports. can be reverted if you want

rbaryanai commented on examples/vnet/doca_dpdk.c:
```diff
@@ -793,6 +812,9 @@ static int doca_dpdk_build_fwd(struct doca_dpdk_pipe *pipe,
 	case DOCA_FWD_RSS:
 		doca_dpdk_build_rss_action(action_entry, fwd_cfg);
 		break;
+    case DOCA_FWD_PORT:
```
here you need to consult with some local cfg object. if doca_dpdk_cfg.nic? then hairpin, else port

salemsol replied:
this is done in doca_dpdk_build_fwd_action. we pass it a bool parameter if true then hairpin.
i added that bool in the fwd_cfg struct that is set upon the init, but can be changed i think this should be in the application domain.
what do you think ?

salemsol committed:
doca-dpdk: Stop incrementing actions counter on end

The action "end" should be the last action in the flow.
currently after adding the end action the actions counter
is wrongly incremented. Fix it

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
doca-dpdk: Fix alignment

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
doca-dpdk: Comment out meter policy calls

rte_mtr_meter_profile_delete and rte_mtr_meter_policy_delete
are creating compilation errors, comment these calls out

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
Include simple_fwd header file

Signed-off-by: Salem Sol <salems@nvidia.com>

salemsol committed:
Introduce simple_fwd application

Signed-off-by: Salem Sol <salems@nvidia.com>
